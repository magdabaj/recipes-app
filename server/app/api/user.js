const { Router } = require('express');
const { hash } = require('../user/helper');
const UserTable = require('../user/table');
const Session = require('../user/session');
const { setSession } = require('./helper');

const router = Router();

router.post('/signup', (req, res, next) => {
  const { email, password } = req.body;

  const emailHash = hash(email);
  const passwordHash = hash(password);

  UserTable.getUser({ emailHash })
    .then(({ user }) => {
      if (!user) {
        return UserTable.storeUser({ emailHash, passwordHash }).catch((error) => next(error));
      } else {
        const error = new Error('This email has already been taken');

        error.statusCode = 409;

        throw error;
      }
    })
    .then((id) => {
      const userId = id.id;
      return setSession({ email, res, userId });
    })
    .then(({ message, userId }) => res.json({ message, userId }))
    .catch((error) => next(error));
});

router.post('/signin', (req, res, next) => {
  const { email, password } = req.body;

  const emailHash = hash(email);
  const passwordHash = hash(password);

  UserTable.getUser({ emailHash })
    .then(({ user }) => {
      if (user && user.passwordHash === passwordHash) {
        const userId = user.id;
        const { sessionId } = user;

        return setSession({ email, res, sessionId, userId });
      } else {
        const error = new Error('Incorrect email or password');

        error.statusCode = 409;

        throw error;
      }
    })
    .then(({ message, userId }) => res.json({ message, userId }))
    .catch((error) => next(error));
});

router.get('/signout', (req, res, next) => {
  const { email } = Session.parse(req.cookies.sessionString);

  UserTable.updateUserId({
    sessionId: null,
    emailHash: hash(email),
  })
    .then(() => {
      res.clearCookie('sessionString');

      res.json({ message: 'Successful logout' });
    })
    .catch((error) => next(error));
});

router.get('/authenticated', (req, res, next) => {
  const { sessionString } = req.cookies;

  if (!sessionString || !Session.verify(sessionString)) {
    const error = new Error('Invalid session');

    error.statusCode = 400;

    return next(error);
  } else {
    const { email, id } = Session.parse(sessionString);

    UserTable.getUser({ emailHash: hash(email) })
      .then(({ user }) => {
        const authenticated = user.sessionId === id;
        const userId = user.id;

        if (authenticated) {
          res.json({ authenticated, email, userId });
        } else {
          res.json({ authenticated });
        }
      })
      .catch((error) => next(error));
  }
});

module.exports = router;
