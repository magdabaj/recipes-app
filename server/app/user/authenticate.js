const Session = require('./session');
const UserTable = require('./table');
const { hash } = require('./helper');

const authenticate = (sessionString, userId) => {
  return new Promise((resolve, reject) => {
    if (!sessionString || !Session.verify(sessionString)) {
      resolve(false);
    } else {
      const { email, id } = Session.parse(sessionString);

      UserTable.getUser({ emailHash: hash(email) })
        .then(({ user }) => {
          let authenticated = user.sessionId === id && user.id === parseInt(userId);

          if (authenticated) {
            resolve({ authenticated, email });
          } else {
            reject(authenticated);
          }
        })
        .catch((error) => {
          return error;
        });
    }
  });
};

module.exports = authenticate;
