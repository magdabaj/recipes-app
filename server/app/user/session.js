const uuid = require('uuid');
const { hash } = require('./helper');

const SEPARATOR = '|';

class Session {
  constructor({ email }) {
    this.email = email;
    this.id = uuid();
  }

  toString() {
    const { email, id } = this;

    return Session.sessionString({ email, id });
  }

  static userData({ email, id }) {
    return `${email}${SEPARATOR}${id}`;
  }

  static sessionString({ email, id }) {
    const userData = Session.userData({ email, id });

    return `${userData}${SEPARATOR}${hash(userData)}`;
  }

  static parse(sessionString) {
    const sessionData = sessionString.split(SEPARATOR);

    return {
      email: sessionData[0],
      id: sessionData[1],
      sessionHash: sessionData[2],
    };
  }

  static verify(sessionString) {
    const { email, id, sessionHash } = Session.parse(sessionString);

    const userData = Session.userData({ email, id });

    return hash(userData) === sessionHash;
  }
}

module.exports = Session;
