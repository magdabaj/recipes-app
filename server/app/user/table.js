const pool = require('../../databasePool');

class UserTable {
  static storeUser({ emailHash, passwordHash }) {
    const addedDate = new Date();
    return new Promise((resolve, reject) => {
      pool.query(
        'INSERT INTO "user"("emailHash", "passwordHash", "addedDate") VALUES ($1,$2, $3) RETURNING id',
        [emailHash, passwordHash, addedDate],
        (error, response) => {
          if (error) return reject(error);

          resolve(response.rows[0]);
        },
      );
    });
  }

  static getUser({ emailHash }) {
    return new Promise((resolve, reject) => {
      pool.query(
        'SELECT id,"passwordHash","sessionId" FROM "user" WHERE "emailHash" = $1',
        [emailHash],
        (error, response) => {
          if (error) return reject(error);

          resolve({ user: response.rows[0] });
        },
      );
    });
  }

  static updateUserId({ sessionId, emailHash }) {
    return new Promise((resolve, reject) => {
      pool.query(
        'UPDATE "user" SET "sessionId" = $1 WHERE "emailHash" = $2',
        [sessionId, emailHash],
        (error, response) => {
          if (error) return reject(error);

          resolve();
        },
      );
    });
  }
}

module.exports = UserTable;
