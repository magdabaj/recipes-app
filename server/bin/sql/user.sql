CREATE TABLE user(
    id SERIAL PRIMARY KEY,
    email CHARACTER (64),
    password CHARACTER (64)
    "addedDate" TIMESTAMP NOT NULL,
);