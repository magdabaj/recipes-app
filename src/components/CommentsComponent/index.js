/**
 *
 * CommentsComponent
 *
 */

import React from 'react';
import '../../containers/RecipeContainer/index.css';
import CommentForm from '../CommentForm';
import { Link } from 'react-router-dom';
import fetchStates from "../../utils/fetchStates";
import { toast } from "react-toastify";
// import PropTypes from 'prop-types';
// import styled from 'styled-components';

function CommentsComponent({
                               comments,
                               addComment,
                               recipeId,
                               user,
                               getComments,
                               editComment,
                               removeComment,
                               ...props
                           }) {
    return (
        <section className={'comments-container'}>
            {user.loggedIn ? (
                <CommentForm
                    addComment={addComment}
                    recipeId={recipeId}
                    user={user}
                    getComments={getComments}
                    {...props}
                />
            ) : (
                <h3>
                    <Link to={'/login'}>Musisz sie zalogowac zeby dodac komentarz</Link>
                </h3>
            )}
            <div>
                {comments.map(comment => (
                    <ol className={'comments-list'}>
                        <li className={'comment-body'}>
                            <article className={'comment-article'}>
                                <footer className={'comment-title'}>
                                    <div className={'comment-author'}>{comment.email}</div>
                                    <div className={'comment-data'}>
                                        {comment.addedDate.replace(/\T/, ' o ').replace(/.Z/g, '')}
                                    </div>
                                </footer>
                                <div className={'comment-content'}>
                                    <p>{comment.content}</p>
                                </div>
                                {comment.email.trim() === user.email ? (
                                    <div className={'comment-edit'}>
                                        <div className={'comment-edit--button'}>Edit</div>
                                        <div
                                            className={'comment-edit--button'}
                                            onClick={() => removeComment(comment.id)}
                                        >
                                            Remove
                                        </div>
                                    </div>
                                ) : null}
                            </article>
                        </li>
                    </ol>
                ))}
            </div>
        </section>
    );
}

CommentsComponent.propTypes = {};

export default CommentsComponent;
