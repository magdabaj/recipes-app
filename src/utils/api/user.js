const api = 'https://afternoon-plateau-23579.herokuapp.com/user';
// const api = 'http://localhost:5000/user'

export const signUpUser = async (email, password) => {
    return fetch(`${api}/signup`, {
        method: 'POST',
        body: JSON.stringify({ email, password }),
        headers: { 'Content-Type': 'application/json' },
        credentials: 'include',
    }).then(res => res.json());
};

export const signInUser = async (email, password) => {
    return fetch(`${api}/signin`, {
        method: 'POST',
        body: JSON.stringify({ email, password }),
        headers: { 'Content-Type': 'application/json' },
        credentials: 'include',
    }).then(res => res.json());
};

export const signOutUser = async () => {
    return fetch(`${api}/signout`, {
        headers: { 'Content-Type': 'application/json' },
        credentials: 'include',
    }).then(res => res.json());
};

export const authenticateUser = async () => {
    return fetch(`${api}/authenticated`, {
        headers: { 'Content-Type': 'application/json' },
        credentials: 'include',
    }).then(res => res.json());
};
