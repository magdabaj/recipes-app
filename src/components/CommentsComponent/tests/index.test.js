import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render } from "@testing-library/react";
import * as jestDOM from '@testing-library/jest-dom';
import CommentsComponent from "../index";
import {fakeUser, loggedOutUser} from "../../../utils/testHelpers/fixtures/user";
import {fakeComments} from "../../../utils/testHelpers/fixtures/coments";
// import commonTests from "../../../utils/testHelpers/commonTests";

expect.extend(jestDOM)
const mockedAddComment = jest.fn()
const removeCommentMocked = jest.fn()
const getCommentsMocked = jest.fn()

test('renders CommentsComponent', () => {
    const { getByLabelText } = render(<CommentsComponent
        user={fakeUser}
        addComment={mockedAddComment}
        recipeId={'1'}
        removeComment={removeCommentMocked}
        getComments={getCommentsMocked}
        comments={fakeComments}
    />)
    const commentForm = getByLabelText(/dodaj komentarz/i)
    expect(commentForm).toBeInTheDocument()
})

test('renders CommentsComponent without comment form', () => {
    const { queryByLabelText, getByRole } = render(<Router>
            <CommentsComponent
            user={loggedOutUser}
            addComment={mockedAddComment}
            recipeId={'1'}
            removeComment={removeCommentMocked}
            getComments={getCommentsMocked}
            comments={fakeComments}
        />
    </Router>)
    const heading = getByRole('heading')

    expect(heading).toBeInTheDocument()
    expect(heading).toHaveTextContent('Musisz sie zalogowac zeby dodac komentarz')
    expect(queryByLabelText(/dodaj komentarz/i)).not.toBeInTheDocument()
})
//
// const renderCommentsComponent = render(<CommentsComponent
//     user={fakeUser}
//     addComment={mockedAddComment}
//     recipeId={'1'}
//     removeComment={removeCommentMocked}
//     getComments={getCommentsMocked}
//     comments={fakeComments}
// />)
//
// commonTests(renderCommentsComponent)