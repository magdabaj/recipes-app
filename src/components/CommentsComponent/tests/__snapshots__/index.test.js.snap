// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render and match the snapshot 1`] = `
<section
  class="comments-container"
>
  <div
    class="comment-form-container"
  >
    <form
      class="comment-form"
    >
      <label
        class="comment-label"
        for="comment-area"
      >
        Dodaj komentarz
      </label>
      <textarea
        class="comment-text-area"
        data-testid="comment-area"
        id="comment-area"
        maxlength="2000"
        name="content"
        placeholder="Wpisz tresc komentarza"
      />
      <button
        class="comment-button"
        type="submit"
      >
        Dodaj komentarz
      </button>
    </form>
  </div>
  <div>
    <ol
      class="comments-list"
    >
      <li
        class="comment-body"
      >
        <article
          class="comment-article"
        >
          <footer
            class="comment-title"
          >
            <div
              class="comment-author"
            >
              test
            </div>
            <div
              class="comment-data"
            >
              test
            </div>
          </footer>
          <div
            class="comment-content"
          >
            <p>
              test
            </p>
          </div>
          <div
            class="comment-edit"
          >
            <div
              class="comment-edit--button"
            >
              Edytuj
            </div>
            <div
              class="comment-edit--button"
              role="remove-comment-button"
            >
              Usuń
            </div>
          </div>
        </article>
      </li>
    </ol>
    <ol
      class="comments-list"
    >
      <li
        class="comment-body"
      >
        <article
          class="comment-article"
        >
          <footer
            class="comment-title"
          >
            <div
              class="comment-author"
            >
              test
            </div>
            <div
              class="comment-data"
            >
              test
            </div>
          </footer>
          <div
            class="comment-content"
          >
            <p>
              test
            </p>
          </div>
          <div
            class="comment-edit"
          >
            <div
              class="comment-edit--button"
            >
              Edytuj
            </div>
            <div
              class="comment-edit--button"
              role="remove-comment-button"
            >
              Usuń
            </div>
          </div>
        </article>
      </li>
    </ol>
  </div>
</section>
`;
